src/librepgp/stream-common.cpp

```diff
@@ -173,62 +172,64 @@ src_peek(pgp_source_t *src, void *buf, size_t len)
         if (src->knownsize && (src->readb + read > src->size)) {
             read = src->size - src->readb;
         }
-        read = src->read(src, &cache->buf[cache->len], read);
-        if (read == 0) {
+        if (!src->read(src, &cache->buf[cache->len], read, &read)) {
+            src->error = 1;
+            return false;
+        }
+        if (!read) {
             if (buf) {
                 memcpy(buf, &cache->buf[0], cache->len);
             }
-            return cache->len;
-        } else if (read < 0) {
-            src->error = 1;
-            return -1;
-        } else {
-            cache->len += read;
-            if (cache->len >= len) {
-                if (buf) {
-                    memcpy(buf, &cache->buf[0], len);
-                }
-                return len;
+            *peeked = cache->len;
+            return true;
+        }
+        cache->len += read;
+        if (cache->len >= len) {
+            if (buf) {
+                memcpy(buf, cache->buf, len);
             }
+            *peeked = len;
+            return true;
         }
     }
+    return false;
+}
 
-    return -1;
+bool
+src_peek_eq(pgp_source_t *src, void *buf, size_t len)
+{
+    size_t res = 0;
+    return src_peek(src, buf, len, &res) && (res == len);
 }
 
-ssize_t
+void
 src_skip(pgp_source_t *src, size_t len)
 {
-    ssize_t res = 0;
-    void *  buf;
-    uint8_t sbuf[16];
-
     if (src->cache && (src->cache->len - src->cache->pos >= len)) {
         src->readb += len;
         src->cache->pos += len;
-        return len;
+        return;
     }
 
+    size_t  res = 0;
+    uint8_t sbuf[16];
     if (len < sizeof(sbuf)) {
-        return src_read(src, sbuf, len);
+        (void) src_read(src, sbuf, len, &res);
+        return;
     }
 
-    buf = calloc(1, std::min((size_t) PGP_INPUT_CACHE_SIZE, len));
+    void *buf = calloc(1, std::min((size_t) PGP_INPUT_CACHE_SIZE, len));
     if (!buf) {
-        return -1;
```

Comment:
@rrrooommmaaa (here and below) src->error is managed in src_read() function, so independent on callback error will be set if callback returns it. Another point - this PR wasn't aimed to change the src->error handling, i.e. anyway it should not change it.

Reply:
@ni4 This piece seems important to me. The previous version returned error code (-1) on memory allocation fault, this version doesn't return any error code, just exits prematurely (before moving the position of src, right?). Hence, we have indeterminate state after calling this function -- the position may or may not have moved. My solution is to put src in error state (at least), so the next read will be guaranteed to fail.

Reply:
@rrrooommmaaa Yeah, it returned -1 which was never checked. So, basically, it would behave in the same way as before. From the 'strict logical' perspective memory allocation failure doesn't relate to the read failure, so it is not 100% correct to set error to 1 here.
And, anyway, on allocation failure all the further code will completely fail eventually.

Reply:
@ni4 How about using a static buffer for skipping? If allocating 32K is too much for a static buffer, how about using a small fallback buffer in case the big buffer can't be alocated. Thus we can guarantee, that the src position will be skipped in any case.

Reply:
... and `src_read` and `src_peek` should exit with error immediately if `src->error` is set. Or would you recommend to do this as a separate PR?

Reply:
@rrrooommmaaa I think it should be filed as a separate issue, since doesn't relate to this PR, and doesn't seem to be of high criticality. src->error is set to just be able to check it later on via src_error() function.